evm/src/cpu/kernel/interpreter.rs

```diff
@@ -326,11 +332,13 @@ impl<'a> Interpreter<'a> {
     }
 
     fn run_prover_input(&mut self) -> anyhow::Result<()> {
-        let input = self
-            .prover_inputs
-            .pop()
-            .ok_or_else(|| anyhow!("Out of prover inputs"))?;
-        self.stack.push(input);
+        let prover_input_fn = self
+            .prover_inputs_map
+            .get(&(self.offset - 1))
+            .ok_or_else(|| anyhow!("Offset not in prover inputs."))?;
+        let output = prover_input_fn.run(self.stack.clone());
```

Comment:
I know using `Vec::pop` could be more convenient, but I think this should probably pass a slice to avoid cloning the stack. Would also help emphasize that it's read-only; at first glance I thought the `pop()`s were actually mutating the stack.